src/erl/typespec.rs

```diff
@@ -0,0 +1,263 @@
+            .intersperse(", ".to_doc()),
+    )
+}
+
+pub fn type_name_to_atom(name: &str) -> String {
+    let name = &name.to_snake_case();
+    if names::is_erlang_reserved_word(name) {
+        format!("'{}'", name)
+    } else {
+        name.to_string()
+    }
+}
+
+/* Generate an erlang '-type' line for this custom type. There are three options for entries:
+
+   1. Basic constructors with no arguments that we convert to a single atom
+
+        -type option(A) :: ... | none
+
+   2. Constructors where all the arguments are labelled, which we treat as records, and output a
+      type which uses the record definition that has already been output and included in this file:
+
+        -type queue(Element) :: #queue{in :: Element, out :: Element} | ..
+
+      We declare and pass through any type parameters so that the type has the right number of
+      arguments and the type parameters are used.
+
+   3. Constructors some or all of the arguments are not labelled, which are output as tag tuples:
+
+        -type option(A) :: {some, A} | ..
+*/
+pub fn type_(name: &str, _parameters: &[String], constructors: &[RecordConstructor]) -> Document {
+    let found_parameters = extract_params_from_constructors(constructors);
+
+    "-type "
+        .to_doc()
+        .append(name.to_snake_case())
+        .append(wrap_args(
+            found_parameters
+                .iter()
+                .map(|name| name.to_camel_case().to_doc()),
+        ))
+        .append(" :: ")
+        .append(concat(
+            constructors
+                .iter()
+                .map(|constructor| {
+                    if constructor.args.is_empty() {
+                        // Basic single-atom constructor
+                        type_name_to_atom(&constructor.name).to_doc()
+                    } else if is_record(&constructor) {
+                        record_type(&constructor)
+                    } else {
+                        non_record_type(&constructor)
+                    }
+                })
+                .intersperse(delim(" |")),
+        ))
+        .append(".")
+}
+
+fn record_type(constructor: &RecordConstructor) -> Document {
+    let name = constructor.name.to_snake_case();
+    let contents = type_record_definition(&constructor.args);
+
+    "#".to_doc().append(name).append(contents)
+}
+
+fn non_record_type(constructor: &RecordConstructor) -> Document {
+    constructor
+        .name
+        .to_snake_case()
+        .to_doc()
+        .append(delim(","))
+        .append(concat(
+            constructor
+                .args
+                .iter()
+                .map(|(_, ast, ..)| type_ast_name(ast, None))
+                .intersperse(delim(",")),
+        ))
+        .surround("{", "}")
+}
+
+fn is_record(constructor: &RecordConstructor) -> bool {
+    for (label, ..) in constructor.args.iter() {
+        match label {
+            Some(_) => return true,
+            None => return false,
+        }
+    }
+
+    false
```

Comment:
Maybe this?

```suggestion
    constructor.args.iter().all(|(label, ..)| label.is_some())
```

Reply:
Nice! Though `.all()` returns true for an empty iterator so we might need:
```rust
    (!constructor.args.is_empty()) && constructor.args.iter().all(|(label, ..)| label.is_some())
```
I'll change it to that.